contracts/upgradeable_contracts/omnibridge_nft/modules/forwarding_rules/MultiTokenForwardingRulesManager.sol

```diff
@@ -0,0 +1,143 @@
+pragma solidity 0.7.5;
+
+import "../OwnableModule.sol";
+
+/**
+ * @title MultiTokenForwardingRulesManager
+ * @dev Multi token mediator functionality for managing destination AMB lanes permissions.
+ */
+contract MultiTokenForwardingRulesManager is OwnableModule {
+    address internal constant ANY_ADDRESS = 0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF;
+
+    // Forwarding rules mapping
+    // token => sender => receiver => destination lane
+    mapping(address => mapping(address => mapping(address => int256))) public forwardingRule;
+
+    event ForwardingRuleUpdated(address token, address sender, address receiver, int256 lane);
+
+    // solhint-disable-next-line no-empty-blocks
+    constructor(address _owner) OwnableModule(_owner) {}
+
+    /**
+     * @dev Tells the destination lane for a particular bridge operation by checking several wildcard forwarding rules.
+     * @param _token address of the token contract on the foreign side of the bridge.
+     * @param _sender address of the tokens sender on the home side of the bridge.
+     * @param _receiver address of the tokens receiver on the foreign side of the bridge.
+     * @return destination lane identifier, where the message should be forwarded to.
+     *  1 - oracle-driven-lane should be used.
+     *  0 - default behaviour should be applied.
+     * -1 - manual lane should be used.
+     */
+    function destinationLane(
```

Comment:
Based on the production experience with the erc20 OB it seems that the following scenario is most needed: we need to forward all collectibles except some specifics to the manual lane. May it be configured with this code?

Reply:
Agree. Then I propose to introduce the following changes to the rules:
* if manager is not configured in the OB contract, all messages are forwarded to the PBO
* if manager is configured in the OB contract, all messages are forwarded to the PBU
* `setTokenForwardingRule(token, true)` allows all operation for the particular token to go to the PBO
* `setSenderExceptionForTokenForwardingRule(token, sender, true)` allows particular sender to bridge particular token using PBO. Makes sense only if `setTokenForwardingRule` was not applied for the token.
* `setReceiverExceptionForTokenForwardingRule(token, receiver, true)` allows particular receiver to bridge particular token using PBO. Makes sense only if `setTokenForwardingRule` was not applied for the token.
* `setSenderForwardingRule(sender, true)` disallow particular sender to use PBO completely.
* `setReceiverForwardingRule(receiver, true)` disallow particular receiver to use PBO completely.

Reply:
> setTokenForwardingRule(token, true) allows all operation for the particular token to go to the PBO
> `setSenderExceptionForTokenForwardingRule(token, sender, true)` allows particular sender to bridge particular token using PBO. Makes sense only if setTokenForwardingRule was not applied for the token.
> `setReceiverExceptionForTokenForwardingRule(token, receiver, true)` allows particular receiver to bridge particular token using PBO. Makes sense only if setTokenForwardingRule was not applied for the token.

So, if a message for a particular token transfer is driven through the PBO lane, does it makes sense to use `setSenderExceptionForTokenForwardingRule`/`setReceiverExceptionForTokenForwardingRule` to forbid some specific sender/receiver to use the PBO and go through request-and-claim instead? For example, we came to a partnership with a project `P` which is responsible for a collectibles `C` and the project agreed to donate the bridge oracles that's why we allowed the collectibles `C` to use the PBO lane. But later we found that there is a person who abuses the oracle by transferring an NFT of the collectibles `C` that's why we would like to send his/her messages through the PBU lane.

> * `setSenderForwardingRule(sender, true)` disallow particular sender to use PBO completely.
> * `setReceiverForwardingRule(receiver, true)` disallow particular receiver to use PBO completely.

If the messages are driving through the PBU lane by default it looks logical to allow some specific accounts to use the PBO lane (e.g. the bridge validators) by using `setSenderForwardingRule`/`setReceiverForwardingRule`, doesn't it?

Reply:
> So, if a message for a particular token transfer is driven through the PBO lane, does it makes sense to use setSenderExceptionForTokenForwardingRule/setReceiverExceptionForTokenForwardingRule to forbid some specific sender/receiver to use the PBO and go through request-and-claim instead? For example, we came to a partnership with a project P which is responsible for a collectibles C and the project agreed to donate the bridge oracles that's why we allowed the collectibles C to use the PBO lane. But later we found that there is a person who abuses the oracle by transferring an NFT of the collectibles C that's why we would like to send his/her messages through the PBU lane.

For that purpose, I left `setSenderForwardingRule`/`setReceiverForwardingRule` to allow restricting particular sender/receiver addresses. 

 `setSenderExceptionForTokenForwardingRule`/`setReceiverExceptionForTokenForwardingRule` are intended to be used for some kind of oracles. I think it might be the case when we want to allow bridging some token through the PBO, only for the particular address.

Maybe it makes sense to rename this methods, if there is some ambiguity.

Reply:
> Maybe it makes sense to rename this methods, if there is some ambiguity.

* if manager is not configured in the OB contract, all messages are forwarded to the PBO
* if manager is configured in the OB contract, all messages are forwarded to the PBU
* `setRuleForTokenToPBO(token, true)` allows all operations for the particular token to go to the PBO
* `setRuleForSenderOfAnyTokenToPBO(sender, true)` allows all operations from the particular sender to go to the PBO
* `setRuleForReceiverOfAnyTokenToPBO(receiver, true)` allows all operations to the particular receiver to go to the PBO
* `setRuleForTokenAndSenderToPBO(token, sender, true)` allows particular sender to bridge particular token using PBO. Makes sense only if `setTokenForwardingRule` was not applied for the token.
* `setRuleForTokenAndReceiverToPBO(token, receiver, true)` allows particular receiver to bridge particular token using PBO. Makes sense only if `setTokenForwardingRule` was not applied for the token.
* `setRuleForSenderOfAnyTokenToPBU(sender, true)` disallow particular sender to use PBO completely.
* `setRuleForReceiverOfAnyTokenToPBU(receiver, true)` disallow particular receiver to use PBO completely.

How about such logic?

Reply:
Looks good to me, I have renamed and added remaining functions as you suggested.